### PR TITLE
plugin/cache: Add refresh mode setting to serve_stale

### PR DESCRIPTION
**Pull request by atollena**
### 1. Why is this pull request needed and what does it do?

This PR adds an optional REFRESH_MODE parameter on the serve_stale configuration directive of the cache plugin. For context, please see #4309

### 2. Which issues (if any) are related?

Reasoning and RFC discussed at #4309
Other similar issues are #4097 and #4199.

### 3. Which documentation changes (if any) need to be made?

I updated the cache plugin README.md to document the new parameter on the PR.

### 4. Does this introduce a backward incompatible change or deprecation?

No.

### CODE REVIEW

**Comment by atollena** [comment]
I'm taking over the work from @Towerthousand on this, started in https://github.com/coredns/coredns/pull/4756. This commit implements suggestions from @chrisohaver:

1. properly return stale records on nxdomain in accordance with rfc 8767
2. rename before/after to immediate/verify

I also refactored the tests slightly to make it easier to test the nxdomain case.

In addition, I think it would be valuable to go a bit further and implement the following bits of rfc8767:

1. Add a failure recheck timer to prevents querying the upstream for a period of time (potentially configurable) -- we have cases where we are worried that when an upstream is down, it gets hammered by clients when coming back. This would help in this scenario.
2. Add a client response timer to respond with stale records early, when a stale record is available. Since we plan to use this for intra-DC DNS, we can set this very low, which reduces latency before getting a stale entry.
3. make the TTL of stale entries  configurable (at the moment it is hardcoded at 0, which the RFC explicitly forbids). Ideally I think we'd go with 30s by default, as recommended by the RFC, but this is a backward incompatible change.

I'd be willing to do any of those items, either directly here or in a follow-up PR, but I'm interested to get feedback before implementing them.

File: plugin/cache/README.md
Comment (chrisohaver) [inline plugin/cache/README.md]: > refresh the cache happens before (`immediate`) or after (`verified`) returning a response

"immediate" and "verified" are swapped here, resulting in incorrect description.
And the phrasing is confusing, probably better described using the option terminology.

* `verified` will first _verify_ that an entry is still unavailable from the source before sending the stale response to the client.
* `immediate` will _immediately_ send a response to the client before checking to see if the entry is available from the source.

File: plugin/cache/handler.go
Comment (chrisohaver) [inline plugin/cache/handler.go]: Calling `c.doRefresh` generally results in sending a response to the client if a response is received - no matter what the response code is. As proposed, if the response code is not NXDOMAIN or NOERROR, the client can be sent two responses, the first one being the current upstream response, and the second one (which would be ignored by the client) being the stale cached entry.

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/cache/README.md
Comment (atollena) [reply plugin/cache/README.md]: fixed, thank you for the suggestion.

File: plugin/cache/handler.go
Comment (atollena) [reply plugin/cache/handler.go]: OK, I realise that I did not understand your comment in https://github.com/coredns/coredns/pull/4756#discussion_r672453981.

I wrapped the response writer as you suggested, so that it only writes if the response is nxdomain or noerror and exposes whether it wrote. I'm not quite sure that the behaviour is correct in case the underlying `ResponseWriter` returns an error to `WriteMsg`. As proposed the code just ignores the error and proceeds with serving stale entry -- my concern is whether this could in certain cases lead to double writing.

**Comment by atollena** [comment]
@chrisohaver thanks for the prompt review, I think I addressed your comments.

File: plugin/cache/README.md
Comment (chrisohaver) [inline plugin/cache/README.md]: Please add a warning about that latency will be higher when using `verified`.  It was present in an earlier commit, but looks like it was removed in this last commit.

File: plugin/cache/cache.go
Comment (chrisohaver) [inline plugin/cache/cache.go]: This should check the response code of the message directly (`res.Rcode == dns.RcodeSuccess || res.Rcode == dns.RcodeNameError`), rather than running the message through `Typify` which returns pseudo-codes (like `NoData`, which has a `NOERROR` return code).

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/cache/README.md
Comment (atollena) [reply plugin/cache/README.md]: I restored the warning (I changed a few words but kept the content).

File: plugin/cache/cache.go
Comment (atollena) [reply plugin/cache/cache.go]: Done.

File: plugin/cache/README.md
Comment (chrisohaver) [inline plugin/cache/README.md]: Latency is always increased for serving stale entries with `verified` ...  not just for unavailable upstreams/sources.  

"can reduce" -> "will prevent": `verified` prevents stale entries from ever being served if an updated response can be retrieved from the source. 

```suggestion
  value to `verified` can lead to increased latency when serving stale responses, but will
  prevent stale entries from ever being served if an updated response can be retrieved from the source.
```

> especially when resolution is infrequent

I think better to remove this qualification. `verified` always prevents this, regardless of resolution frequency.  Conversely, `immediate` will serve  1 extra stale result to a client before updating the cache.  If there is high concurrency on the cache item, then many of those stale responses can be written to clients before the cache is updated.

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Code LGTM.  Just some minor tweaks to the readme now.

File: plugin/cache/README.md
Comment (atollena) [reply plugin/cache/README.md]: I took your suggestion.

**Review by chrisohaver — APPROVED** [review]

**Comment by atollena** [comment]
I realised by looking at the coverage report that I didn't properly test discarding messages that are not success or nxdomain, so I improved the test a bit to write a response in case of servfail.

**Comment by atollena** [comment]
Thanks again @chrisohaver for the quick review rounds. Is there anything I can or should do to help get this change merged? (e.g. squashing the git commits, or anything else)

**Comment by chrisohaver** [comment]
> Thanks again @chrisohaver for the quick review rounds. Is there anything I can or should do to help get this change merged?

Unless a PR is really trivial, I usually defer final approval and merging to a codeowner of the related plugin. In doing reviews of code that I am not a codeowner for, my hope is that this effort makes their reviews easier and PRs merge more quickly.

**Comment by atollena** [comment]
IIUC @miekg is the sole owner of the cache plugin. Miek, do you have any feedback for this change?

I'm also interested to have your input on [this comment above](https://github.com/coredns/coredns/pull/5131#issuecomment-1023208795) regarding further work on serve_stale behaviors -- let me know if you'd rather have that filed as an issue for further discussion.

**Comment by chrisohaver** [comment]
I'm going to merge this in the next couple of days unless anyone objects.

**Comment by atollena** [comment]
> I'm going to merge this in the next couple of days unless anyone objects.

Thanks. I merged master and squashed the commits.

**Review by chrisohaver — APPROVED** [review]

File: plugin/cache/README.md
Comment (lobshunter) [inline plugin/cache/README.md:61]: A minor typo here, `verified` will cause a invalid value error

```suggestion
  the cache happens. `verify` will first verify that an entry is still unavailable from the source before sending
```

Comment (chrisohaver) [reply plugin/cache/README.md:61]: @lobshunter, This PR merged a couple months ago, and the typo has since been fixed.

Comment (lobshunter) [reply plugin/cache/README.md:61]: My bad, I checked out to this PR to see how it works, didn't notice the new fix. 😂

**Comment by salla2** [comment]
> I'm taking over the work from @Towerthousand on this, started in #4756. This commit implements suggestions from @chrisohaver:
> 
> 1. properly return stale records on nxdomain in accordance with rfc 8767
> 2. rename before/after to immediate/verify
> 
> I also refactored the tests slightly to make it easier to test the nxdomain case.
> 
> In addition, I think it would be valuable to go a bit further and implement the following bits of rfc8767:
> 
> 1. Add a failure recheck timer to prevents querying the upstream for a period of time (potentially configurable) -- we have cases where we are worried that when an upstream is down, it gets hammered by clients when coming back. This would help in this scenario.
> 2. Add a client response timer to respond with stale records early, when a stale record is available. Since we plan to use this for intra-DC DNS, we can set this very low, which reduces latency before getting a stale entry.
> 3. make the TTL of stale entries  configurable (at the moment it is hardcoded at 0, which the RFC explicitly forbids). Ideally I think we'd go with 30s by default, as recommended by the RFC, but this is a backward incompatible change.
> 
> I'd be willing to do any of those items, either directly here or in a follow-up PR, but I'm interested to get feedback before implementing them.

@atollena Did you tried implementing above changes? especially failure recheck timer to prevents querying the upstream for a period of time ?

**Comment by atollena** [comment]
> > I'm taking over the work from @Towerthousand on this, started in #4756. This commit implements suggestions from @chrisohaver:
> > 
> > 1. properly return stale records on nxdomain in accordance with rfc 8767
> > 2. rename before/after to immediate/verify
> > 
> > I also refactored the tests slightly to make it easier to test the nxdomain case.
> > In addition, I think it would be valuable to go a bit further and implement the following bits of rfc8767:
> > 
> > 1. Add a failure recheck timer to prevents querying the upstream for a period of time (potentially configurable) -- we have cases where we are worried that when an upstream is down, it gets hammered by clients when coming back. This would help in this scenario.
> > 2. Add a client response timer to respond with stale records early, when a stale record is available. Since we plan to use this for intra-DC DNS, we can set this very low, which reduces latency before getting a stale entry.
> > 3. make the TTL of stale entries  configurable (at the moment it is hardcoded at 0, which the RFC explicitly forbids). Ideally I think we'd go with 30s by default, as recommended by the RFC, but this is a backward incompatible change.
> > 
> > I'd be willing to do any of those items, either directly here or in a follow-up PR, but I'm interested to get feedback before implementing them.
> 
> @atollena Did you tried implementing above changes? especially failure recheck timer to prevents querying the upstream for a period of time ?

I did not and do not plan to work on this further. Feel free to take that over.

**Pull request merged**
